src/glob.rs

```diff
@@ -528,44 +669,272 @@ mod test {
+        run_file_patterns_match_test(&file_patterns, path, kind, expected);
+      };
+    run_test(&cwd, PathKind::Directory, FilePatternsMatch::Passed);
+    run_test(
+      &cwd.join("other"),
+      PathKind::Directory,
+      FilePatternsMatch::Passed,
+    );
+    run_test(
+      &cwd.join("other/sub_dir"),
+      PathKind::Directory,
+      FilePatternsMatch::Passed,
+    );
+    run_test(
+      &cwd.join("not_matched"),
+      PathKind::File,
+      FilePatternsMatch::Excluded,
+    );
+    run_test(
+      &cwd.join("other/test.ts"),
+      PathKind::File,
+      FilePatternsMatch::Passed,
+    );
+    run_test(
+      &cwd.join("other/test.js"),
+      PathKind::File,
+      FilePatternsMatch::Excluded,
+    );
+  }
+
+  #[test]
+  fn file_patterns_exclude() {
+    let cwd = std::env::current_dir().unwrap();
+    let file_patterns = FilePatterns {
+      base: cwd.clone(),
+      include: None,
+      exclude: PathOrPatternSet(vec![
+        PathOrPattern::from_relative(&cwd, "target").unwrap(),
+        PathOrPattern::from_relative(&cwd, "!not_excluded").unwrap(),
+        // lower items take priority
+        PathOrPattern::from_relative(&cwd, "excluded_then_not_excluded")
+          .unwrap(),
+        PathOrPattern::from_relative(&cwd, "!excluded_then_not_excluded")
+          .unwrap(),
+        PathOrPattern::from_relative(&cwd, "!not_excluded_then_excluded")
+          .unwrap(),
+        PathOrPattern::from_relative(&cwd, "not_excluded_then_excluded")
+          .unwrap(),
+      ]),
+    };
+    let run_test =
+      |path: &Path, kind: PathKind, expected: FilePatternsMatch| {
+        run_file_patterns_match_test(&file_patterns, path, kind, expected);
+      };
+    run_test(&cwd, PathKind::Directory, FilePatternsMatch::Passed);
+    run_test(
+      &cwd.join("target"),
+      PathKind::File,
+      FilePatternsMatch::Excluded,
+    );
+    run_test(
+      &cwd.join("not_excluded"),
+      PathKind::File,
+      FilePatternsMatch::PassedOptedOutExclude,
+    );
+    run_test(
+      &cwd.join("excluded_then_not_excluded"),
+      PathKind::File,
+      FilePatternsMatch::PassedOptedOutExclude,
+    );
+    run_test(
+      &cwd.join("not_excluded_then_excluded"),
+      PathKind::File,
+      FilePatternsMatch::Excluded,
+    );
+  }
+
+  #[test]
+  fn file_patterns_include_exclude() {
+    let cwd = std::env::current_dir().unwrap();
+    let file_patterns = FilePatterns {
+      base: cwd.clone(),
+      include: Some(PathOrPatternSet(vec![
+        PathOrPattern::from_relative(&cwd, "other").unwrap(),
+        PathOrPattern::from_relative(&cwd, "target").unwrap(),
+        PathOrPattern::from_relative(&cwd, "**/*.js").unwrap(),
+        PathOrPattern::from_relative(&cwd, "**/file.ts").unwrap(),
+      ])),
+      exclude: PathOrPatternSet(vec![
+        PathOrPattern::from_relative(&cwd, "target").unwrap(),
+        PathOrPattern::from_relative(&cwd, "!target/other/**").unwrap(),
```

Comment:
We're gonna need some solid docs on this new functionality, negated exclude takes a while to figure out